File: src/main/scala/com/vc/medium/L950.scala
package com.vc.medium

object L950 {
  def deckRevealedIncreasing(deck: Array[Int]): Array[Int] = {
    import scala.collection.mutable
    val n = deck.length

    java.util.Arrays.sort(deck)

    val q = new mutable.Queue[Int]()
    (0 until n).foreach(i => q.enqueue(i))

    var i = 0
    val res = new Array[Int](n)
    while(i < n) {
      res(q.dequeue) = deck(i)
      if(q.nonEmpty)q.enqueue(q.dequeue)
      i += 1
    }
    res
  }
}

File: src/main/java/com/vc/medium/L950.java
package com.vc.medium;

import java.util.*;

class L950 {
    public int[] deckRevealedIncreasing(int[] deck) {
        if(deck == null || deck.length == 0) return new int[0];

        int n = deck.length;

        Queue<Integer> q = new LinkedList<>();
        for(int i = 0; i < n; i++) q.offer(i);

        Arrays.sort(deck);
        int[] res = new int[n];
        int index = 0;
        while(!q.isEmpty()) {
            //Reveal
            int reveal = q.poll();
            res[reveal] = deck[index++];

            //Push to the bottom
            if(!q.isEmpty()) q.offer(q.poll());
        }
        return res;
    }
}
